README.md
```bash
  _____ _    _ _____  ______ _____  _      _______   __
 / ____| |  | |  __ \|  ____|  __ \| |    / ____\ \ / /
| (___ | |  | | |__) | |__  | |__) | |   | |     \ V / 
 \___ \| |  | |  ___/|  __| |  _  /| |   | |      > <  
 ____) | |__| | |    | |____| | \ \| |___| |____ / . \ 
|_____/ \____/|_|    |______|_|  \_\______\_____/_/ \_\
```

<div align="center">
  <a href="https://travis-ci.com/github/dashjay/superlcx"><img src="https://travis-ci.com/dashjay/superlcx.svg?branch=master" alt="Build Status"></a>
  <a href="https://github.com/dashjay/superlcx/actions?query=workflow%3AAutoRelease"><img src="https://github.com/dashjay/superlcx/workflows/AutoRelease/badge.svg" alt="Release Status"></a>
  <a href="https://github.com/dashjay/superlcx/actions?query=workflow%3ABuild"><img src="https://github.com/dashjay/superlcx/workflows/Build/badge.svg" alt="Build Status"></a>
  <a href="https://github.com/dashjay/superlcx/actions?query=workflow%3ATest"><img src="https://github.com/dashjay/superlcx/workflows/Test/badge.svg" alt="Test Status"></a>
</div>

[EN][中文](./README.CN.md)

# intro
A high performance tool with some middleware. SuperLcx ACTS as a proxy for the request and response value from the server. During the process, some middleware will be called to realize some operations:
- Modify the request (forward request).
- Modify the response (sub_filter).
- Log requests and response (dump traffics).
- Hook operations set by yourself, such as implementing a Python hook through GRPC, or changing the behavior of middleware through an interpreted language such as Lua or Js.
- implement your own middleware...

# usage
```bash
Usage of ./superlcx:
  -M string
        middleware, comma separated if more than one, eg: --M stdout,dumps
  -c string
        load config from
  -host string
        target host:port. (default "0.0.0.0:8081")
  -l int
        listen port (default 8080)
  -log string
        l -> line of code, d -> date, t -> time, order doesn't matter (default "t")
  -m string
        run mode <proxy|copy|blend>. (default "proxy")
  -pp int
        pprof port (default 8999)
  -v    show version and about then exit.
```

### mode
- proxy 
    - advantages: the proxy mode work with http proxy package which can offer(expose) more api like modifyResponse,Transport,director, etc.
    - disadvantages: the proxy mode will cause memory jitter, not suitable for limited memory machine if you need high performance

- copy
    - advantages: the copy mode directly work on tcp layer, it doesn't care about what would be transfer. With the help of `io.Copy`, it needs less RAM.
    - disadvantages: the copy mode know nothing about application layer. all things it can do is dumping them all out.
    
- blend
    - advantages: allocate low memory comparing with the proxy mode. it can run with the middleware interface.
    - disadvantages: the blend mode still need more memory(less than proxy), and could lead to memory jitter.

### -c Configuration file
> read configured with a Golang [toml library](https://github.com/BurntSushi/toml)

⚠ If you read the configuration file with -c, please write all the configuration files, do not rely partly on the command line parameters, we do not know the internal implementation of the TOML library, may overwrite the origin config.

For the documentation of the configuration file, please see: [Configuration Description](./docs/config.md)


### -M middleware
When working in the proxy mode, middleware can be invoked to analyze the traffic in the process. For example, the built-in stdout middleware (sample middleware) can be used via '-M stdout'.
(Must be in the proxy or blend mode to work)

If you want to implement your own middleware, check out:
[Middleware standard](./docs/middleware.md)

Once every thing ok, it can be placed under the middlewares folder. It is disambiguated tha that the file structure remain as follows.
```
middlewares
└── stdout
    └── handler.go
```

The interfaces are exposing as follows under 'handler.go'. If you need to load the configuration, load it yourself, or use the cc module TOML to load your config.
```
func HandleRequest(req *http.Request)
func HandleResponse(req *http.Response)
```

**WARNING:** Middleware components may do something on the req and resp body, That's bound to do a lot of io with bufio, the memory jitters can be very serious.
I thought about writing a center middleware to dump the req and resp out, and then call all middleware like pipeline, but if some of them change the req or resp, it can also cause other problems, so now middleware components do not affect each other.

CHANGELOG.txt
v1.0.6:
    1. add unit test, and adjust some config, making sap stable.

    1. 添加单元测试，并且调整配置，使得程序更加可靠。

v1.0.5:
    1. use upx for compress binaries.

    1. 使用 upx 对二进制程序进行压缩。

v1.0.4:
    1. add js_lua middleware.

    1. 添加js_lua 中间件。

v1.0.3:
    1. add middleware sub_filter.

    1. 添加中间件sub_filter.

v1.0.2:
    1. add config module
    2. add c_header middleware for custom header of req and resp

    1. 添加配置读取文件
    2. 添加c_header中间件，可以自定义请求头和返回头

v1.0.1:
    1. add blendmode.

    1. 添加混合模式

v1.0.0:
    1. add middleware mode for superlcx.

    1. 为superlcx添加中间件模式。

cmd/main.go
package main

import (
	"context"
	"flag"
	"fmt"
	"log"
	"net/http"
	_ "net/http/pprof"
	"os"
	"os/signal"
	"regexp"
	"runtime"
	"strconv"
	"strings"
	"syscall"
	"time"

	. "superlcx/cc"
	"superlcx/core"
)

const version = "1.0.5"

var (
	showVersion bool
	configFile  string
)

func init() {
	defer func() {
		err := recover()
		if err != nil {
			fmt.Print(err)
			flag.PrintDefaults()
			os.Exit(-1)
		}
	}()
	flag.BoolVar(&showVersion, "v", false, "show version and about then exit.")
	flag.StringVar(&configFile, "c", "", "load config from")
	flag.IntVar(&Config.ListenPort, "l", 8080, "listen port")
	flag.IntVar(&Config.PPROFPort, "pp", 8999, "pprof port")
	flag.StringVar(&Config.DefaultTarget, "host", "0.0.0.0:8081", "target host:port.")
	flag.StringVar(&Config.Mode, "m", "proxy", "run mode <proxy|copy|blend>.")
	flag.StringVar(&Config.Middleware, "M", "", "middleware, comma separated if more than one, eg: --M stdout,dumps")
	flag.StringVar(&Config.LogFlag, "log", "t", "l -> line of code, d -> date, t -> time, order doesn't matter")
	flag.Parse()
	if showVersion {
		fmt.Printf(`
  _____ _    _ _____  ______ _____  _      _______   __
 / ____| |  | |  __ \|  ____|  __ \| |    / ____\ \ / /
| (___ | |  | | |__) | |__  | |__) | |   | |     \ V / 
 \___ \| |  | |  ___/|  __| |  _  /| |   | |      > <  
 ____) | |__| | |    | |____| | \ \| |___| |____ / . \ 
|_____/ \____/|_|    |______|_|  \_\______\_____/_/ \_\

Superlcx [%s], a tool kit for port transfer with middlewares!
`, version)
		os.Exit(0)
	}

	if configFile != "" {
		err := Config.InitConfig(configFile)
		if err != nil {
			panic(err)
		}
	}

	if Config.ListenPort < 1 || Config.ListenPort > 65535 {
		panic("[x] Listen Port Invalid")
	}
	checkHost(Config.DefaultTarget)
}

func main() {
	// Buried point for debug
	go http.ListenAndServe(fmt.Sprintf(":%d", Config.PPROFPort), nil)
	go showMemLog()
	ctx, cancel := context.WithCancel(context.Background())

	var (
		// sigs, when anything unexpected happened, a signal will send to
		// this chan. then server start to stop.
		sigs = make(chan os.Signal, 1)
		// when everything closed, a signal will send to done. the main Goroutine then stop.
		done = make(chan bool, 1)
	)
	go func() {
		switch Config.Mode {
		case "proxy":
			c := core.NewSapProxy()
			c.Serve(ctx)
		case "copy":
			c := core.NewSapCopy()
			c.Serve(ctx)
		case "blend":
			c := core.NewSapBlend()
			c.Serve(ctx)
		default:
			flag.PrintDefaults()
			sigs <- os.Interrupt
		}
	}()
	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
	go func() {
		<-sigs
		cancel()
		done <- true
	}()
	<-done
}

// checkHost check the ip:port valid
func checkHost(host string) {
	const pattern = `^(\d{1,2}|1\d\d|2[0-4]\d|25[0-5])\.(\d{1,2}|1\d\d|2[0-4]\d|25[0-5])\.(\d{1,2}|1\d\d|2[0-4]\d|25[0-5])\.(\d{1,2}|1\d\d|2[0-4]\d|25[0-5])$`
	ipPort := strings.Split(host, ":")
	if len(ipPort) != 2 {
		panic("host should like this ip:port ")
	}
	port, err := strconv.Atoi(ipPort[1])
	if err != nil {
		panic(err)
	}
	if port < 1 || port > 65535 {
		panic(fmt.Sprintf("host port %d invalid", port))
	}
	ok, err := regexp.MatchString(pattern, ipPort[0])
	if err != nil || !ok {
		panic(fmt.Sprintf("host ip %s invalid", ipPort[0]))
	}
}

func showMemLog() {
	ticker := time.NewTicker(20 * time.Second)
	for range ticker.C {
		var m runtime.MemStats
		runtime.ReadMemStats(&m)
		log.Printf("Memory: Alloc = %vMb TotalAlloc = %vMb Sys = %vMb NumGC = %v", m.Alloc/1024/1024, m.TotalAlloc/1024/1024, m.Sys/1024/1024, m.NumGC)
	}
}
